pkg/pipeline/controller.go

```diff
@@ -264,128 +266,111 @@ func (c *Controller) UpdateStream(ctx context.Context, req *livekit.UpdateStream
 		return errors.ErrNonStreamingPipeline
 	}
 
-	sendUpdate := false
 	errs := errors.ErrArray{}
-	now := time.Now().UnixNano()
+	sendUpdate := false
 
 	// add stream outputs first
 	for _, rawUrl := range req.AddOutputUrls {
 		// validate and redact url
-		url, redacted, err := o.ValidateUrl(rawUrl, o.OutputType)
+		stream, err := o.AddStream(rawUrl, o.OutputType)
 		if err != nil {
 			errs.AppendErr(err)
 			continue
 		}
 
 		// add stream info to results
 		c.mu.Lock()
-		streamInfo := &livekit.StreamInfo{
-			Url:    redacted,
-			Status: livekit.StreamInfo_ACTIVE,
-		}
-		o.StreamInfo[url] = streamInfo
-
-		c.Info.StreamResults = append(c.Info.StreamResults, streamInfo)
+		c.Info.StreamResults = append(c.Info.StreamResults, stream.StreamInfo)
 		if list := (*livekit.EgressInfo)(c.Info).GetStream(); list != nil {
-			list.Info = append(list.Info, streamInfo)
+			list.Info = append(list.Info, stream.StreamInfo)
 		}
 		c.mu.Unlock()
 
 		// add stream
-		if err = c.streamBin.AddStream(url); err != nil {
-			streamInfo.Status = livekit.StreamInfo_FAILED
+		sendUpdate = true
+		if err = c.streamBin.AddStream(stream); err != nil {
+			stream.StreamInfo.Status = livekit.StreamInfo_FAILED
 			errs.AppendErr(err)
 			continue
 		}
 
-		if o.OutputType != types.OutputTypeRTMP {
-			streamInfo.StartedAt = now
-		}
-		c.OutputCount++
-		sendUpdate = true
+		c.OutputCount.Inc()
 	}
 
 	// remove stream outputs
 	for _, rawUrl := range req.RemoveOutputUrls {
-		url, err := o.GetStreamUrl(rawUrl)
+		stream, err := o.GetStream(rawUrl)
 		if err != nil {
 			errs.AppendErr(err)
 			continue
 		}
 
-		if err = c.removeSink(ctx, url, nil); err != nil {
+		sendUpdate = true
+		if err = c.streamFinished(ctx, stream); err != nil {
 			errs.AppendErr(err)
-		} else {
-			sendUpdate = true
 		}
 	}
 
 	if sendUpdate {
 		c.Info.UpdatedAt = time.Now().UnixNano()
-		_, _ = c.ipcServiceClient.HandlerUpdate(ctx, (*livekit.EgressInfo)(c.Info))
+		c.streamUpdates(func() {
```

Comment:
All updated will be delayed by at least 500ms. Could this cause temporary odd state mismatches when accessing the stream state after updating the egress?

Reply:
it should grab the info and update time at the time it's executed, so even if that state isn't the same as when this was queued up it should always be accurate/up to date